chain_of_trust.go

```diff
@@ -0,0 +1,172 @@
+	TaskID      string                       `json:"taskId"`
+	RunID       uint                         `json:"runId"`
+	WorkerGroup string                       `json:"workerGroup"`
+	WorkerID    string                       `json:"workerId"`
+	Extra       CotExtra                     `json:"extra"`
+}
+
+func (cot *ChainOfTrust) Initialise() error {
+	return nil
+}
+
+func (cot *ChainOfTrust) IsEnabled(fl EnabledFeatures) bool {
+	return fl.GenerateCertificate
+}
+
+func (cot *ChainOfTrust) Killed(task *TaskRun) error {
+	logFile := filepath.Join(TaskUser.HomeDir, "public", "logs", "live_backing.log")
+	certifiedLogFile := filepath.Join(TaskUser.HomeDir, "public", "logs", "certified.log")
+	signedCert := filepath.Join(TaskUser.HomeDir, "public", "logs", "certified.log.gpg")
+	err := appendFileContents(logFile, certifiedLogFile)
+	if err != nil {
+		return err
+	}
+	err = task.uploadLog("public/logs/certified.log")
+	if err != nil {
+		return err
+	}
+	artifactHashes := []ArtifactHash{}
+	for _, artifact := range task.Artifacts {
+		switch a := artifact.(type) {
+		case S3Artifact:
+			// make sure SHA256 is calculated
+			hash, err := calculateHash(a)
+			if err != nil {
+				return err
+			}
+			artifactHashes = append(
+				artifactHashes,
+				ArtifactHash{
+					Name: a.CanonicalPath,
+					Hash: "sha256:" + hash,
+				})
+		}
+	}
+
+	cotCert := &ChainOfTrustCertificate{
+		Artifacts:   artifactHashes,
+		Task:        task.Definition,
+		TaskID:      task.TaskID,
+		RunID:       task.RunID,
+		WorkerGroup: config.WorkerGroup,
+		WorkerID:    config.WorkerID,
+		Extra: CotExtra{
+			PublicIPAddress:  config.PublicIP.String(),
+			PrivateIPAddress: config.PrivateIP.String(),
+			InstanceID:       config.InstanceID,
+			InstanceType:     config.InstanceType,
+			Region:           config.Region,
+		},
+	}
+
+	certBytes, err := json.MarshalIndent(cotCert, "", "  ")
+	if err != nil {
+		return err
+	}
+	// separate signature from json with a new line
+	certBytes = append(certBytes, '\n')
+
+	in := bytes.NewBuffer(certBytes)
+	out, err := os.Create(signedCert)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	out.Write(certBytes)
+
+	privKeyFile, err := os.Open(config.SigningKeyLocation)
+	if err != nil {
+		return err
+	}
+	defer privKeyFile.Close()
+	entityList, err := openpgp.ReadArmoredKeyRing(privKeyFile)
+	if err != nil {
+		return err
+	}
+	privKey := entityList[0]
+	openpgp.ArmoredDetachSign(out, privKey, in, nil)
+	out.Close()
+	err = task.uploadLog("public/logs/certified.log.gpg")
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (cot *ChainOfTrust) Created(task *TaskRun) error {
+	return nil
+}
+
+func appendFileContents(src, dst string) (err error) {
+	in, err := os.Open(src)
+	if err != nil {
+		return
+	}
+	defer in.Close()
+	out, err := os.OpenFile(dst, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
+	if err != nil {
+		return
+	}
+	defer func() {
+		cerr := out.Close()
+		if err == nil {
```

Comment:
No, this is called after function has completed. If `err` is not `nil`, it will be returned. However, if it is `nil`, then `cerr` will be returned, which may be `nil` or not `nil`. This is a way to be able to return the error from `out.Close()` even though it is in a `defer` block.